SMLHelper/Handlers/CoordinatedSpawnsHandler.cs

```diff
@@ -220,6 +220,40 @@ public SpawnInfo(string classId, Vector3 spawnPosition, Vector3 rotation)
             spawnType = SpawnType.ClassId;
         }
 
+        [JsonConstructor]
+        internal SpawnInfo(string classId, TechType techType, Vector3 spawnPosition, Vector3 rotation)
+        {
+            this.classId = classId;
+            this.techType = techType;
+            this.spawnPosition = spawnPosition;
+            this.rotation = Quaternion.Euler(rotation);
+            spawnType = this.techType == TechType.None ? SpawnType.ClassId : SpawnType.TechType;
+        }
+
+        /// <summary>
+        /// Checks if the passed object is equal to the current object
+        /// </summary>
+        /// <param name="obj">passed object</param>
+        /// <returns>true if equal; otherwise false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (obj is not SpawnInfo spawnInfo)
+                return false;
+
+            // this was necessary because the default Equality check was always giving a false positive.
+            return spawnInfo.classId == this.classId && spawnInfo.rotation == this.rotation && spawnInfo.techType == this.techType && spawnInfo.spawnType == this.spawnType;
+        }
+
+        /// <summary>
+        /// Serves as the default hash function.
+        /// </summary>
+        /// <returns>A hash code for the current object.</returns>
+        public override int GetHashCode()
+        {
+            // this was required for C# to shut up
+            return base.GetHashCode();
+        }
```

Comment:
Your `Equals` and `GetHashCode` need some work. I'm concerned that `spawnInfo.rotation == this.rotation` may now in some cases through false negatives, but this might be an acceptable caveat.
However, the `GetHashCode` definitely needs an implementation that reflects the new `Equals` method.
I'll see what I can do on this.

Reply:
As of commit baa7a47 I have implemented what I believe to be an appropriate `Equals` and `GetHashCode` combination for this class.

## Builds baa7a47
### Subnautica
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768011/SMLHelper_SN.STABLE_v2.10.0.2.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768012/SMLHelper_SN.EXP_v2.10.0.2.zip)

### Below Zero
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768014/SMLHelper_BZ.STABLE_v2.10.0.2.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768017/SMLHelper_BZ.EXP_v2.10.0.2.zip)

Reply:
## Builds 5555948
### Subnautica
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768155/SMLHelper_SN.STABLE_v2.10.0.3.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768157/SMLHelper_SN.EXP_v2.10.0.3.zip)

### Below Zero
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768159/SMLHelper_BZ.STABLE_v2.10.0.3.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768161/SMLHelper_BZ.EXP_v2.10.0.3.zip)

Reply:
It is worth noting that once these `Equals(object b)` and `GetHashCode()` are added to the `public SpawnInfo` class, *any changes to these methods will be a breaking change*, as any consumers of the SMLHelper API using `SpawnInfo` in for example a `HashSet` will suddenly have unexpected results. The implications of this need to be considered very carefully before this can be released.

Reply:
As an alternative to overriding `GetHashCode()` and `Equals(object obj)`, I would propose that we instead implement `IEquatable<SpawnInfo>` instead, overriding `Equals(SpawnInfo other)`. This does not require us to implement `GetHashCode()`, and only changes the behaviour of `Equals(SpawnInfo other)` to be based on value equality, whereas `Equals(object obj)` will remain based on reference equality, as is the standard for a C# class. This will mean that `HashSet`s operate as you expect.

For the sake of value-based equality, other approaches I considered but rejected are:
- Convert `SpawnInfo` to a `struct`. This would be a breaking change, as it will change the fundamentals how `Equals` operates, similar to the original proposed change by @Metious.
- Convert `SpawnInfo` to a `record`. Unfortunately, it is not possible with a `record` to override `Equals`, and the default implementation will not give the same results as desired by @Metious, so this will not work, and also will be a breaking change.

Implementing `IEquatable<SpawnInfo>` is the least egregious version of these implementations, but only if it provides everything that is required to fix the bug where List.Contains() is returning false after serialization & deserialization. As such, I am attaching builds to this comment with this implementation for testing by @Metious and @LeeTwentyThree.

## Builds 7267bc4
### Subnautica
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768478/SMLHelper_SN.STABLE_v2.10.0.4.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768486/SMLHelper_SN.EXP_v2.10.0.4.zip)

### Below Zero
[Stable](https://github.com/SubnauticaModding/SMLHelper/files/6768489/SMLHelper_BZ.STABLE_v2.10.0.4.zip)
[Experimental](https://github.com/SubnauticaModding/SMLHelper/files/6768491/SMLHelper_BZ.EXP_v2.10.0.4.zip)

Reply:
Upon further discussion with @Metious, it seems that the `SpawnInfo` class is not in widespread use, so a redesign of the class is unlikely to have a big impact.

As such, converting it to a `struct` so that value-based equality is the expected default makes sense and is acceptable.

Attached is a build that mixes the best of v2.10.0.3 and v.2.10.0.5 builds, implementing the same `Equals(object obj)` and `GetHashCode` of v2.10.0.3, but also implementing `IEquatable<SpawnInfo>` and `Equals(SpawnInfo other)`, and on top of that, also implementing overloads for the `==` and `!=` operators, to keep everything consistent.

Since the `SpawnInfo` class has been changed to a `struct`, I also took the opportunity to fix the naming violations of its `internal` properties. Attached are builds for v2.10.0.6 as of commit b9802ce.

## Builds b9802ce
### Subnautica
[Stable v2.10.0.6](https://github.com/SubnauticaModding/SMLHelper/files/6770222/SMLHelper_SN.STABLE_v2.10.0.6.zip)
[Experimental v2.10.0.6](https://github.com/SubnauticaModding/SMLHelper/files/6770224/SMLHelper_SN.EXP_v2.10.0.6.zip)

### Below Zero
[Stable v2.10.0.6](https://github.com/SubnauticaModding/SMLHelper/files/6770230/SMLHelper_BZ.STABLE_v2.10.0.6.zip)
[Experimental v2.10.0.6](https://github.com/SubnauticaModding/SMLHelper/files/6770231/SMLHelper_BZ.EXP_v2.10.0.6.zip)